library/core/src/ptr/mod.rs

```diff
@@ -547,6 +585,78 @@ pub const fn invalid_mut<T>(addr: usize) -> *mut T {
     addr as *mut T
 }
 
+/// Convert an address back to a pointer, picking up a previously 'exposed' provenance.
+///
+/// This is equivalent to `addr as *const T`. The provenance of the returned pointer is that of *any*
+/// pointer that was previously passed to [`expose_addr`][pointer::expose_addr] or a `ptr as usize`
+/// cast. If there is no previously 'exposed' provenance that justifies the way this pointer will be
```

Comment:
I wonder if we want to weaken the wording here, such as "any pointer that was passed to expose_addr _by this crate_".

That wording might let us add more runtime security on permissive-provenance CHERI in the future — we could additionally shield all crates from each other at runtime with [polyfill #1](https://hackmd.io/I-zHe9bZTJOY78orP1aYIA?view#Appendix-CHERI-polyfill-2-lower-overhead-still-more-secure-than-non-CHERI).

Reply:
The main point of these functions is to support legacy code that can't work with `addr` and `with_addr`. I don't think we can just impose constraints like this on such code.

Note that this is very fragile, e.g. consider a crate defining a newtype like `NonNull` that wraps raw pointers and all their operations -- then all 'expose'ing would seem to come from that same crate, even though it's actually acting "on behalf of" some other crate.